src/geometry/euclidean.lean

```diff
@@ -645,4 +647,40 @@ begin
                                                 (λ he, h2 ((vsub_eq_zero_iff_eq V).1 he))
 end
 
+/-- The inner product of two vectors given with `weighted_vsub`, in
+terms of the pairwise distances. -/
+lemma inner_weighted_vsub {ι₁ : Type*} {s₁ : finset ι₁} {w₁ : ι₁ → ℝ} (p₁ : ι₁ → P)
+    (h₁ : ∑ i in s₁, w₁ i = 0) {ι₂ : Type*} {s₂ : finset ι₂} {w₂ : ι₂ → ℝ} (p₂ : ι₂ → P)
+    (h₂ : ∑ i in s₂, w₂ i = 0) :
+  inner (s₁.weighted_vsub V p₁ w₁) (s₂.weighted_vsub V p₂ w₂) =
+    (-∑ i₁ in s₁, ∑ i₂ in s₂,
+      w₁ i₁ * w₂ i₂ * (dist (p₁ i₁) (p₂ i₂) * dist (p₁ i₁) (p₂ i₂))) / 2 :=
+begin
+  erw [finset.weighted_vsub_of_point_apply, finset.weighted_vsub_of_point_apply,
+       inner_sum_smul_sum_smul_of_sum_eq_zero _ h₁ _ h₂],
```

Comment:
I think this `erw` indicates you're missing a lemma about `s₁.weighted_vsub V p₁ w₁`.

Reply:
I've added and used `weighted_vsub_apply` (with a comment noting that in most cases `weighted_vsub_eq_weighted_vsub_of_point_of_sum_eq_zero` and `weighted_vsub_of_point_apply` would be used instead) to avoid the `erw` here.

Apart from using `erw` or adding a lemma, another approach in such cases of wanting a rewrite to see through a definition is to `unfold` that definition before the rewrite. Is adding a lemma (possibly proved with `rfl`) generally preferred over either of the other approaches?

Reply:
If it's reasonable to mark the lemma with `@[simp]` --- particularly if these simp lemmas cover the all the ways you can "exit" the given definition (e.g. by applying to an argument, taking a projection, etc), then that's much preferable to manually using `unfold` or `dsimp`, because it gives everyone else an easy tool to use afterwards, without having to know anything about your definitions.

I'm pretty maximalist about `simp` lemmas. I want `simp` to work like this:

> The first analogy that came to my mind is of
immersing the nut in some softening liquid, and why not simply
water? From time to time you rub so the liquid penetrates better,
and otherwise you let time pass. The shell becomes more flexible
through weeks and months—when the time is ripe, hand pressure
is enough, the shell opens like a perfectly ripened avocado!

(Grothendieck, inevitably)